src/Compilers/Shared/BuildServerConnection.cs

```
@@ -118,47 +118,62 @@ internal sealed class BuildServerConnection
 
-                            if (!holdsMutex)
-                            {
-                                return new RejectedBuildResponse();
-                            }
-                        }
-                        catch (AbandonedMutexException)
+                        if (!holdsMutex)
                         {
-                            holdsMutex = true;
+                            return new RejectedBuildResponse();
                         }
                     }
-
-                    // Check for an already running server
-                    var serverMutexName = GetServerMutexName(pipeName);
-                    bool wasServerRunning = WasServerMutexOpen(serverMutexName);
-                    var timeout = wasServerRunning ? timeoutExistingProcess : timeoutNewProcess;
-
-                    if (wasServerRunning || tryCreateServerFunc(clientDir, pipeName))
+                    catch (AbandonedMutexException)
                     {
-                        pipeTask = TryConnectToServerAsync(pipeName, timeout, cancellationToken);
+                        holdsMutex = true;
                     }
                 }
-                finally
+
+                // Check for an already running server
+                var serverMutexName = GetServerMutexName(pipeName);
+                bool wasServerRunning = WasServerMutexOpen(serverMutexName);
+                var timeout = wasServerRunning ? timeoutExistingProcess : timeoutNewProcess;
+
+                if (wasServerRunning || tryCreateServerFunc(clientDir, pipeName))
+                {
+                    pipeTask = TryConnectToServerAsync(pipeName, timeout, cancellationToken);
+                }
+            }
+            finally
+            {
+                if (clientMutex != null)
                 {
                     if (holdsMutex)
                     {
                         clientMutex.ReleaseMutex();
```

>clientMutex.ReleaseMutex(); [](start = 24, length = 27)

Can this throw? Should we try/catch around it?

---

It can throw but only in cases where there is a programming error: called from wrong thread for example. Want those cases to throw vs. swallowing an error here.